byzcoin/contract_name.go

```diff
@@ -0,0 +1,286 @@
+	// contract is not controlled by its own darc. Instead, it is
+	// controlled by the darc that guards the instance ID in the invoke
+	// argument.
+
+	// Check the number of signers match with the number of signatures.
+	if len(inst.SignerIdentities) != len(inst.Signatures) {
+		return errors.New("lengh of identities does not match the length of signatures")
+	}
+	if len(inst.Signatures) == 0 {
+		return errors.New("no signatures - nothing to verify")
+	}
+
+	// Check the signature counters.
+	if err := verifySignerCounters(rst, inst.SignerCounter, inst.SignerIdentities); err != nil {
+		return err
+	}
+
+	// Get the darc, we have to do it differently than the normal
+	// verification because the darc that we are interested in is the darc
+	// that guards the instance ID in the instruction.
+	if inst.Invoke == nil {
+		// TODO this needs to be changed when we add delete
+		return errors.New("only invoke is supported")
+	}
+	value := inst.Invoke.Args.Search("instanceID")
+	if value == nil {
+		return errors.New("argument instanceID is missing")
+	}
+	_, _, cID, dID, err := rst.GetValues(value)
+	if err != nil {
+		return err
+	}
+	d, err := LoadDarcFromTrie(rst, dID)
+	if err != nil {
+		return err
+	}
+
+	// Check that the darc has the right permission to allow naming.
+	action := "_name:" + cID
+	ex := d.Rules.Get(darc.Action(action))
+	if len(ex) == 0 {
+		return fmt.Errorf("action '%v' does not exist", action)
+	}
+
+	// Save the identities that provide good signatures.
+	goodIdentities := make([]string, 0)
+	for i := range inst.Signatures {
+		if err := inst.SignerIdentities[i].Verify(msg, inst.Signatures[i]); err == nil {
+			goodIdentities = append(goodIdentities, inst.SignerIdentities[i].String())
+		}
+	}
+	if len(goodIdentities) == 0 {
+		return errors.New("all signatures failed to verify")
+	}
+
+	// Evaluate the expression using the good signatures.
+	getDarc := func(str string, latest bool) *darc.Darc {
+		if len(str) < 5 || string(str[0:5]) != "darc:" {
+			return nil
+		}
+		darcID, err := hex.DecodeString(str[5:])
+		if err != nil {
+			return nil
+		}
+		d, err := LoadDarcFromTrie(rst, darcID)
+		if err != nil {
+			return nil
+		}
+		return d
+	}
+	return darc.EvalExpr(ex, getDarc, goodIdentities...)
+}
+
+func (c *contractNaming) Spawn(rst ReadOnlyStateTrie, inst Instruction, coins []Coin) (sc []StateChange, cout []Coin, err error) {
+	cout = coins
+	var buf []byte
+	// For the very first pointer, we use the default InstanceID value.
+	buf, err = protobuf.Encode(&ContractNamingBody{Latest: InstanceID{}})
+	if err != nil {
+		return
+	}
+	sc = []StateChange{
+		// We do not need a darc ID because the verification works
+		// differently than normal contracts. See the verification
+		// function for more details.
+		NewStateChange(Create, NamingInstanceID, ContractNamingID, buf, nil),
+	}
+	return
+}
+
+type contractNamingEntry struct {
+	IID     InstanceID
+	Prev    InstanceID
```

Comment:
It looks the same for the `Latest` field of the contract body. Are you planning to use those for a later implementation, or is it useful for the client ? I have a hard time to understand what is their purpose from the code itself even if I remember our talk vaguely.

Reply:
It's useful for looking up all the named instances later on so that we do not have any "dangling" ones. It is more useful for the admin because the admin could mark the unused ones as removed. Indeed, we don't have an admin API for that yet, but there is a test to check that this reversed linked list works. `Latest` is basically the latest one in the linked list.